### ISSUE TITLE
Client bundle contains server deps

### ISSUE THREAD

**Issue by ducan-ne** [lxsmnsyc/thaler#2]
I got this when use thaler + vite:
<img width="880" alt="CleanShot 2023-04-07 at 23 31 39@2x" src="https://user-images.githubusercontent.com/16853175/230644215-407e6be4-3d0c-4893-b96a-35ff09445e91.png">

After deep dive to the compiler, I noticed that imports for RPC did not remove from client bundle, it just like a import and unused like below
<img width="723" alt="CleanShot 2023-04-07 at 23 33 34@2x" src="https://user-images.githubusercontent.com/16853175/230644427-7e438e12-5f07-497e-8a87-b81b96af22fc.png">

I don't think this is thaler's job (Vite should removes them), but have no idea how to fix it 😿

**Comment by ducan-ne** [comment]
it seems a issue with puppeteer-core + playwright-core, tested another package and gets no issue.
going to close this issue

**Comment by ducan-ne** [comment]
it seems an issue with cjs

**Comment by lxsmnsyc** [comment]
Yeah I'm thinking about removing unused imports but I'm not sure it's the right way to do since it may ruin the order of side-effects, hence why my initial decision is to delegate this problem to the bundler for treeshaking.

I'll revisit this again in the future. I'm reopening this for now in case someone submits a similiar issue

**Comment by ducan-ne** [comment]
@lxsmnsyc FYI it actually includes server code, but somehow it didn't throw an error when I build the code
In the following code, it shipped aws-sdk s3 to client
<img width="1305" alt="CleanShot 2023-04-14 at 22 22 46@2x" src="https://user-images.githubusercontent.com/16853175/232086612-609e8dd6-5189-4680-a9a9-0b2bc5b7fd72.png">

For who faced this issue, a workaround could be use `vite-plugin-external` to externalize your deps to `window.XXXX`

**Comment by ducan-ne** [comment]
I think we can take a look on next.js source code to know their approach. I didn't get any problem with next.js originally

**Comment by lxsmnsyc** [comment]
@ducan-ne based on what me and some people have tested, the usual effective approach here is to dynamically import the server module (in the server function) rather than relying on the top-level import declaration. This is a bit manual work but should cover most if not all of the challenges.

The thing with top-level declarations is that tree-shaking will not guarantee the removal of the import since some module are side-effects (hence my motivation to not tamper with import declarations). The intent for the import is never clear.
